aten/src/ATen/native/cuda/IndexKernel.cu

```diff
@@ -24,53 +25,23 @@ static constexpr int launch_bound2 = 4;
 
 static constexpr int launch_size_nd = 128;
 
-template <int Dims, typename T, typename IndexType>
-__device__ __forceinline__ IndexType indexToOffset(
-    const cuda::detail::TensorInfo<T, IndexType>& info,
+template <typename scalar_t, typename index_t>
+__device__ __forceinline__ index_t indexToOffset(
+    const cuda::detail::TensorInfo<scalar_t, index_t>& info,
     int64_t index,
-    IndexType size) {
-  IndexType linearIndex = static_cast<IndexType>(index);
-  CUDA_KERNEL_ASSERT(linearIndex < size && linearIndex >= -size);
+    index_t size,
+    const bool contiguous=false) {
+  index_t linearIndex = static_cast<index_t>(index);
   if (linearIndex < 0) {
     linearIndex += size;
   }
-  return cuda::detail::IndexToOffset<T, IndexType, Dims>::get(linearIndex, info);
-}
-
-template<typename IndexType, typename T>
-void dispatchTakePutImpl(const Tensor& input, Tensor& output, const Tensor& index) {
-  auto inputInfo = cuda::detail::getTensorInfo<T, IndexType>(input);
-  inputInfo.collapseDims();
-  auto numel = input.numel();
-  if (inputInfo.isContiguous()) {
-    cuda::CUDA_tensor_apply2<T, int64_t>(
-        output,
-        index,
-        [inputInfo, numel] __device__ (
-            T & out, const int64_t& idx) {
-            auto offset = indexToOffset<-2, T, IndexType>(inputInfo, idx, numel);
-            out = inputInfo.data[offset];
-        });
+  if (contiguous) {
+    return linearIndex;
   } else {
-    cuda::CUDA_tensor_apply2<T, int64_t>(
-        output,
-        index,
-        [inputInfo, numel] __device__ (
-            T & out, const int64_t& idx) {
-            auto offset = indexToOffset<-1, T, IndexType>(inputInfo, idx, numel);
-            out = inputInfo.data[offset];
-        });
+    return cuda::detail::IndexToOffset<scalar_t, index_t, -1>::get(linearIndex, info);
```

Comment:
do you have benchmarks for non-contiguous tensor? Usually this `-1` specialization is very slow.

Reply:
No I don't but it is the one that was used before. At first sight, I do not see how could it be done any better, as you do the index computation once and only once for every element of the `idx` tensor. Do you know how could it be improved?

Reply:
Usually it has a  few hardcoded instantiations like
```
if (input.ndim() == 1) {
    return cuda::detail::IndexToOffset<scalar_t, index_t, 1>::get(linearIndex, info)
} else if (input.ndim()==2) {
   .....
} else {
   ...instantiate general case with -1
}
```
but if the perf is not awful, it's not needed (I've just looked up, and e.g. indexAddSmallIndex is instantiated for 1,2,3-dimensional inputs

Reply:
I unrolled it. I get about a 5% improvement in the sizes I tested. It is not very much, but the code is not much more complicated really, just a few extra calls. What do you think? If you think I should go for this, I can also do a similar trick on CPU, so that the compiler can unroll the loop.